Source/WebKit/UIProcess/API/wpe/WebKitInputMethodContextImplWPE.cpp

```diff
@@ -0,0 +1,202 @@
+
+                *underlines = g_list_prepend(*underlines, underline);
+            }
+            g_list_free_full(wpeUnderlines, reinterpret_cast<GDestroyNotify>(wpe_input_method_underline_free));
+        }
+    }
+
+    if (cursorOffset)
+        *cursorOffset = clampTo<unsigned>(offset);
+}
+
+static void webkitInputMethodContextImplWPENotifyFocusIn(WebKitInputMethodContext* context)
+{
+    auto* priv = WEBKIT_INPUT_METHOD_CONTEXT_IMPL_WPE(context)->priv;
+    wpe_input_method_context_focus_in(priv->context.get());
+}
+
+static void webkitInputMethodContextImplWPENotifyFocusOut(WebKitInputMethodContext* context)
+{
+    auto* priv = WEBKIT_INPUT_METHOD_CONTEXT_IMPL_WPE(context)->priv;
+    wpe_input_method_context_focus_out(priv->context.get());
+}
+
+static void webkitInputMethodContextImplWPENotifyCursorArea(WebKitInputMethodContext* context, int x, int y, int width, int height)
+{
+    auto* priv = WEBKIT_INPUT_METHOD_CONTEXT_IMPL_WPE(context)->priv;
+    wpe_input_method_context_set_cursor_area(priv->context.get(), x, y, width, height);
+}
+
+static void webkitInputMethodContextImplWPENotifySurrounding(WebKitInputMethodContext* context, const gchar* text, unsigned length, unsigned cursorIndex, unsigned selectionIndex)
+{
+    auto* priv = WEBKIT_INPUT_METHOD_CONTEXT_IMPL_WPE(context)->priv;
+    wpe_input_method_context_set_surrounding(priv->context.get(), text, length, cursorIndex, selectionIndex);
+}
+
+static void webkitInputMethodContextImplWPEReset(WebKitInputMethodContext* context)
+{
+    auto* priv = WEBKIT_INPUT_METHOD_CONTEXT_IMPL_WPE(context)->priv;
+    wpe_input_method_context_reset(priv->context.get());
+}
+
+static void webkit_input_method_context_impl_wpe_class_init(WebKitInputMethodContextImplWPEClass* klass)
+{
+    auto* imClass = WEBKIT_INPUT_METHOD_CONTEXT_CLASS(klass);
+    imClass->get_preedit = webkitInputMethodContextImplWPEGetPreedit;
+    imClass->notify_focus_in = webkitInputMethodContextImplWPENotifyFocusIn;
+    imClass->notify_focus_out = webkitInputMethodContextImplWPENotifyFocusOut;
+    imClass->notify_cursor_area = webkitInputMethodContextImplWPENotifyCursorArea;
+    imClass->notify_surrounding = webkitInputMethodContextImplWPENotifySurrounding;
+    imClass->reset = webkitInputMethodContextImplWPEReset;
+}
+
+WebKitInputMethodContext* webkitInputMethodContextImplWPENew(WPEView *view)
+{
+    auto* context = WEBKIT_INPUT_METHOD_CONTEXT(g_object_new(WEBKIT_TYPE_INPUT_METHOD_CONTEXT_IMPL_WPE, nullptr));
+
+    g_signal_connect_swapped(context, "notify::input-purpose", G_CALLBACK(inputPurposeChangedCallback), context);
+    g_signal_connect_swapped(context, "notify::input-hints", G_CALLBACK(inputHintsChangedCallback), context);
+
+    auto* priv = WEBKIT_INPUT_METHOD_CONTEXT_IMPL_WPE(context)->priv;
+    priv->context = adoptGRef(wpe_input_method_context_new(wpe_view_get_display(view)));
+    wpe_input_method_context_set_client_view(priv->context.get(), view);
```

Comment:
I'm not sure we need this on wpe since we always set it on construction and we never set it to NULL.

Reply:
If I understand this now correctly, you are suggesting that view instance should be passed in wpe_input_method_context_new instead of passing display? Meaning changing line 190 as follows
```diff
-priv->context = adoptGRef(wpe_input_method_context_new(wpe_view_get_display(view)));
+priv->context = adoptGRef(wpe_input_method_context_new(view));
```